Source/WebCore/Modules/mediastream/RTCDataChannel.cpp

```diff
@@ -248,10 +254,15 @@ void RTCDataChannel::didDetectError(Ref<RTCError>&& error)
 void RTCDataChannel::bufferedAmountIsDecreasing(size_t amount)
 {
     queueTaskKeepingObjectAlive(*this, TaskSource::Networking, [this, amount] {
+#if USE(LIBWEBRTC)
         auto previousBufferedAmount = m_bufferedAmount;
         m_bufferedAmount -= amount;
         if (previousBufferedAmount > m_bufferedAmountLowThreshold && m_bufferedAmount <= m_bufferedAmountLowThreshold)
             dispatchEvent(Event::create(eventNames().bufferedamountlowEvent, Event::CanBubble::No, Event::IsCancelable::No));
+#else
+        UNUSED_VARIABLE(amount);
+        dispatchEvent(Event::create(eventNames().bufferedamountlowEvent, Event::CanBubble::No, Event::IsCancelable::No));
```

Comment:
It seems that what you are after is something like bufferedAmountIsBelowThreshold.
It might be clearer to introduce that one instead of reusing bufferedAmountIsDecreasing with a dummy amount.

I also wonder how you will be able to implement the spec, which is based on a bufferedAmount which is manipulated on the context's thread, and not the backend background thread.
For instance, if the backend hops to the main thread, they need to call queueTaskKeepingObjectAlive.
Between the time queueTaskKeepingObjectAlive is called and the time the event is fired, the web app might have called send multiple times, which will have updated m_bufferedAmount.
How will you cancel firing the event in that case?

Reply:
You're right, this is not going to be spec-compliant. I really need to have GStreamer pass the `amount` back to us. I'll revert this part of the patch.

Reply:
I do not see how GstWebRTC can avoid the racy behavior, given we call queueTaskKeepingObjectAlive which will dispatch the event asynchronously.
If we want to avoid the raciness, we would need GstWebRTC to compute the buffered amount in the JS thread, and dispatch the event synchronously without the queueTaskKeepingObjectAlive.
Maybe all this can be moved to GstWebRTC specific code (GStreamerDataChannelHandler.cpp).

Also m_bufferedAmount would never be decreased in your implementation, which is observable from JS (RTCDataChannel as a bufferedAmount getter). Firing the event with a bad bufferedAmount does not seem great.

> I doubt a patch will be accepted wiring up an `amount` value from GstWebRTC to the app because it's not part of the public spec API surface.

It is: https://w3c.github.io/webrtc-pc/#dom-datachannel-bufferedamount

Reply:
If GstWebRTC has a way to get bufferedAmount, it could be monitor this way.
Or WebKit GstWebRTC class could try to generate the low memory signal if it knows when a message has been sent?

Reply:
Yes, GstWebRTCDataChannel has a `buffered-amount` property which we can monitor. The `amount` I was referring to previously is the delta between new amount and previous amount. Is that exposed in the spec? I thought it wasn't.

I'll update the patch and try another approach for events dispatching...